### PR TITLE
Upgrade tonic and prost

### LINKED ISSUE #274: RUSTSEC-2020-0016: `net2` crate has been deprecated; use `socket2` instead
> `net2` crate has been deprecated; use `socket2` instead

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Status              | unmaintained                |
| Package             | `net2`                      |
| Version             | `0.2.37`                   |
| URL                 | [https://github.com/deprecrated/net2-rs/commit/3350e3819adf151709047e93f25583a5df681091](https://github.com/deprecrated/net2-rs/commit/3350e3819adf151709047e93f25583a5df681091) |
| Date                | 2020-05-01                         |

The [`net2`](https://crates.io/crates/net2) crate has been deprecated
and users are encouraged to considered [`socket2`](https://crates.io/crates/socket2) instead.

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2020-0016.html) for additional details.

### LINKED ISSUE #275: RUSTSEC-2020-0056: stdweb is unmaintained
> stdweb is unmaintained

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Status              | unmaintained                |
| Package             | `stdweb`                      |
| Version             | `0.4.20`                   |
| URL                 | [https://github.com/koute/stdweb/issues/403](https://github.com/koute/stdweb/issues/403) |
| Date                | 2020-05-04                         |

The author of the `stdweb` crate is unresponsive.

Maintained alternatives:

- [`wasm-bindgen`](https://github.com/rustwasm/wasm-bindgen)
- [`js-sys`](https://github.com/rustwasm/wasm-bindgen/tree/master/crates/js-sys)
- [`web-sys`](https://github.com/rustwasm/wasm-bindgen/tree/master/crates/web-sys)

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2020-0056.html) for additional details.

### LINKED ISSUE #423: RUSTSEC-2021-0124: Data race when sending and receiving after closing a `oneshot` channel
> Data race when sending and receiving after closing a `oneshot` channel

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Package             | `tokio`                      |
| Version             | `0.1.22`                   |
| URL                 | [https://github.com/tokio-rs/tokio/issues/4225](https://github.com/tokio-rs/tokio/issues/4225) |
| Date                | 2021-11-16                         |
| Patched versions    | `>=1.8.4, <1.9.0,>=1.13.1`                  |
| Unaffected versions | `<0.1.14`               |

If a `tokio::sync::oneshot` channel is closed (via the
[`oneshot::Receiver::close`] method), a data race may occur if the
`oneshot::Sender::send` method is called while the corresponding
`oneshot::Receiver` is `await`ed or calling `try_recv`.

When these methods are called concurrently on a closed channel, the two halves
of the channel can concurrently access a shared memory location, resulting in a
data race. This has been observed to [cause memory corruption][corruption].

Note that the race only occurs when **both** halves of the channel are used
after the `Receiver` half has called `close`. Code where `close` is not used, or where the
`Receiver` is not `await`ed and `try_recv` is not called after calling `close`,
is not affected.

See [tokio#4225][issue] for more details.

[corruption]: https://github.com/tokio-rs/tokio/issues/4225#issuecomment-967434847
[issue]: https://github.com/tokio-rs/tokio/issues/4225
[`oneshot::Receiver::close`]: https://docs.rs/tokio/1.14.0/tokio/sync/oneshot/struct.Receiver.html#method.close

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2021-0124.html) for additional details.

### LINKED ISSUE #587: RUSTSEC-2023-0005: `tokio::io::ReadHalf<T>::unsplit` is Unsound
> `tokio::io::ReadHalf&lt;T&gt;::unsplit` is Unsound

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Status              | unsound                |
| Package             | `tokio`                      |
| Version             | `0.2.25`                   |
| URL                 | [https://github.com/tokio-rs/tokio/issues/5372](https://github.com/tokio-rs/tokio/issues/5372) |
| Date                | 2023-01-11                         |

`tokio::io::ReadHalf&lt;T&gt;::unsplit` can violate the `Pin` contract

The soundness issue is described in the [tokio/issues#5372](https://github.com/tokio-rs/tokio/issues/5372)

Specific set of conditions needed to trigger an issue (a !Unpin type in ReadHalf)
is unusual, combined with the difficulty of making any arbitrary use-after-free
exploitable in Rust without doing a lot of careful alignment of data types in
the surrounding code.

The `tokio` feature `io-util` is also required to be enabled to trigger this
soundness issue.

Thanks to zachs18 reporting the issue to Tokio team responsibly and taiki-e
and carllerche appropriately responding and fixing the soundness bug.

Tokio before 0.2.0 used `futures` 0.1 that did not have `Pin`, so it is not
affected by this issue.

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2023-0005.html) for additional details.

### LINKED ISSUE #589: RUSTSEC-2023-0034: Resource exhaustion vulnerability in h2 may lead to Denial of Service (DoS)
> Resource exhaustion vulnerability in h2 may lead to Denial of Service (DoS)

| Details             |                                                |
| ------------------- | ---------------------------------------------- |
| Package             | `h2`                      |
| Version             | `0.2.7`                   |
| URL                 | [https://github.com/hyperium/hyper/issues/2877](https://github.com/hyperium/hyper/issues/2877) |
| Date                | 2023-04-14                         |
| Patched versions    | `>=0.3.17`                  |

If an attacker is able to flood the network with pairs of `HEADERS`/`RST_STREAM` frames, such that the `h2` application is not able to accept them faster than the bytes are received, the pending accept queue can grow in memory usage. Being able to do this consistently can result in excessive memory use, and eventually trigger Out Of Memory.

This flaw is corrected in [hyperium/h2#668](https://github.com/hyperium/h2/pull/668), which restricts remote reset stream count by default.

See [advisory page](https://rustsec.org/advisories/RUSTSEC-2023-0034.html) for additional details.

### PR DESCRIPTION
**Pull request by diconico07**
**What this PR does / why we need it**:
This PR initial goal is to upgrade tonic and prost so we can drop the build dependency to rustfmt, thus making it easier to re-build akri in an air-gaped system.

However, to attain this goal it was necessary to do two other unrelated changes (more details about each one after):
 - Upgrade h2
 - Upgrade to edition 2021 of rust

### Upgrade h2 
The only really needed part for tonic and prost is removing the pin to a patched h2 `0.3.3`, however for consistency I also upgraded actix (used in webhook part).
This solves several RUSTSEC issues (some of them already marked as resolved for akri, but a cargo audit still complained):
- Fixes #274 
- Fixes #275 
- Fixes #423
- Fixes #589 
- Fixes #587 

### Upgrade to edition 2021 of rust 
Prost now generates code using 2021 edition of rust, so we need to upgrade this in order to upgrade prost.

**Special notes for your reviewer**:
This might be a hard to read PR, I split it in several commits that should be a bit easier to parse (at least they only upgrade one thing)

**If applicable**:
- [ ] this PR has an associated PR with documentation in [akri-docs](https://github.com/project-akri/akri-docs)
- [ ] this PR contains unit tests
- [x] added code adheres to standard Rust formatting (`cargo fmt`)
- [x] code builds properly (`cargo build`)
- [x] code is free of common mistakes (`cargo clippy`)
- [x] all Akri tests succeed (`cargo test`)
- [x] inline documentation builds (`cargo doc`)
- [x] all commits pass the [DCO bot check](https://probot.github.io/apps/dco/) by being signed off -- see the failing DCO check for instructions on how to retroactively sign commits

### CODE REVIEW

**Comment by diconico07** [comment]
This PR depends on #594 as a change in intermediate containers is needed

File: .github/workflows/check-rust.yml
Comment (jbpaux) [inline .github/workflows/check-rust.yml:1]: Shouldn't we also delete this part ?
https://github.com/project-akri/akri/blob/1ec427b271369b37147dff75c056a56b2bd36628/.github/workflows/check-rust.yml#L52-L55

Comment (kate-goldenring) [reply .github/workflows/check-rust.yml:1]: +1

File: Cargo.toml [hunk 1]
```diff
@@ -1,8 +1,3 @@
-# Patch to allow invalid authority headers provided by grpc-go / kubelet
-# Issue to track: https://github.com/grpc/grpc-go/issues/2628
-[patch.crates-io]
```
Comment (kate-goldenring) [inline Cargo.toml:3]: YAHOOOO! So happy to see this mess cleaned up

File: webhooks/validating/configuration/src/main.rs [hunk 2]
```diff
@@ -677,7 +677,7 @@ mod tests {
     }
 
     #[test]
-    fn test_deserialized_has_extra_array_element() {
+    async fn test_deserialized_has_extra_array_element() {
```
Comment (kate-goldenring) [inline webhooks/validating/configuration/src/main.rs:680]: Why are all of these changed to `async`? They all look synchronous and are using the `#[test]` macro still.

Comment (diconico07) [reply webhooks/validating/configuration/src/main.rs]: `actix_web` now defines a `test` macro that got pulled here, went to fast during upgrade and thought it was already the case, fixed it now.

**Comment by kate-goldenring** [comment]
@diconico07 not sure why the cross production code workflow is failing. Doesn't seem to be an issue with rust version compat https://crates.io/crates/cross

**Comment by diconico07** [comment]
Okay, seems like we're going to need a way more recent protobuf-compiler package in the cross image than the one package in ubuntu 14.04, and upgrading to latest cross image will only takes us to 18.04 that don't have recent enough protobuf-compiler either, their next release should go to 20.04 that will have a recent enough version

**Comment by diconico07** [comment]
In fact the rustembedded one was the official one back then, but they moved to ghcr since.
The latest release from them (0.2.5) uses 18.04 with protobuf compiler in version 2.6.1 (not enough for us).
They should soon release a new one that is based on 20.04, we could use the "main" image for now to unblock this, but I think it's better to wait for the next cross release.

**Comment by jbpaux** [comment]
Indeed, I wasn't aware of this renaming 😅. I saw that too. Not sure when 0.3.0 will come out, soon I guess as they are already 2 months past due date, just 9 issues to close.

**Comment by diconico07** [comment]
They (cross) moved their estimated due date, and still late on it, I'm more and more enticed to drop cross entirely, I'll raise the topic on the slack to gather some thoughts on this.

**Comment by diconico07** [comment]
Rebased on #670 to make it build correctly, and upgraded all dependencies (with a focus on tonic and prost).

With that upgrade, the h2 patch is still needed (using the one from hyperium/h2#612, but as agreed with @kate-goldenring, we are going to maintain our own fork to use in Cargo.toml to ensure a better supply chain while we try to get this upstream).

Another patch is needed for OPCUA crate, this patch is only the main upstream branch, as we need an upstream not yet released fix to work with latest rust version (some checks are now stricter)

**Comment by kate-goldenring** [comment]
> Another patch is needed for OPCUA crate, this patch is only the main upstream branch, as we need an upstream not yet released fix to work with latest rust version (some checks are now stricter)

@diconico07 which change/ pr in the `opcua` crate are we waiting to be released?  Looks like it still hasn't been released since the pr points to v12 when that hasn't been released yet https://github.com/locka99/opcua

**Comment by diconico07** [comment]
> > Another patch is needed for OPCUA crate, this patch is only the main upstream branch, as we need an upstream not yet released fix to work with latest rust version (some checks are now stricter)
> 
> @diconico07 which change/ pr in the `opcua` crate are we waiting to be released? Looks like it still hasn't been released since the pr points to v12 when that hasn't been released yet https://github.com/locka99/opcua

This is the merged PR we are waiting: https://github.com/locka99/opcua/pull/260

**Comment by kate-goldenring** [comment]
@diconico07 maybe we take this as a sign to move the opcua DH out too?

**Comment by diconico07** [comment]
@kate-goldenring probably, but I'd wait for udev one to be moved before doing this. It hopefully means this override is going to be short-lived in this repository.

**Comment by kate-goldenring** [comment]
Looks like `opcua` has a new release: https://github.com/locka99/opcua/releases/tag/0.12.0. I think we can update to use this release and be ready to merge

**Review by kate-goldenring — APPROVED** [review]

**Pull request merged**